DESIGN.md

```diff
@@ -0,0 +1,756 @@
+
+Because io-uring differs significantly from epoll, Tokio must provide a new set
+of APIs to take full advantage of the reduced overhead. However, Tokio's
+[stability guarantee][stability] means Tokio APIs cannot change until 2024 at
+the earliest. Additionally, the io-uring API is still evolving with [new
+functionality][tweet] planned for the near future. Instead of waiting for
+io-uring to mature and a Tokio 2.0 release, we will release a standalone crate
+dedicated to exposing an io-uring optimal API. This new crate will be able to
+iterate rapidly with breaking changes without violating Tokio's stability
+guarantee. Applications deployed exclusively on Linux kernels 5.10 or later may
+choose to use this crate when taking full advantage of io-uring's benefits
+provides measurable benefits. Examples of intended use-cases include TCP
+proxies, HTTP file servers, and databases.
+
+[overhead]: https://www.usenix.org/system/files/conference/nsdi14/nsdi14-paper-jeong.pdf
+[bench]: https://github.com/frevib/io_uring-echo-server/blob/master/benchmarks/benchmarks.md
+[stability]: https://tokio.rs/blog/2020-12-tokio-1-0#a-stability-guarantee
+[tweet]: https://twitter.com/axboe/status/1371978266806919168
+
+# Guide-level explanation
+
+The tokio-uring crate provides a module layout similar to the Tokio crate,
+including modules such as net (TCP, UDP, and Unix domain sockets), fs (file
+system access), io (standard in, out, and error streams). It also provides a
+runtime module containing an io-uring specific runtime. Modules such as sync and
+any other containing async Rust utilities are not included by tokio-uring, in
+favor of the ones provided by the main Tokio crate.
+
+```toml
+[dependencies]
+tokio-uring = "0.1"
+```
+
+```rust
+fn main() {
+    let rt = tokio_uring::runtime::Runtime::new().unwrap();
+    rt.block_on(async {
+        // The rest of the application comes here.
+    });
+}
+```
+
+The application's `main` function starts a tokio-uring runtime and launches its
+asynchronous logic within that. The tokio-uring runtime can drive both io-uring
+specific resources (e.g., `TcpStream` and `TcpListener`) and Tokio resources,
+enabling any library written for Tokio to run within the io-uring specific
+runtime.
+
+## Submit-based operations
+
+Operations on io-uring backed resources return futures, representing the
+operation completion. The caller awaits the future to get the operation result.
+
+```rust
+let socket = my_listener.accept().await?;
+```
+
+The runtime communicates with the kernel using two single-producer,
+single-consumer queues. It submits operation requests, such as accepting a TCP
+socket, to the kernel using the submission queue. The kernel then performs the
+operation. On completion, the kernel returns the operation results via the
+completion queue and notifies the process. The `io_uring_enter` syscall flushes
+the submission queue and acquires any pending completion events. Upon request,
```

Comment:
What does it mean for a specific syscall to "acquire a completion event" ?  (i.e. The syscall itself cannot own responsibility for handling that event, right? So is it handing off that responsibility to whomever in user space invoked the syscall ?)

Reply:
(Or, I guess since the syscall is allowed to block the thread, it *can* have that responsibility? But below you say waiting for a *minimum* number of completion events; not *all* of the events... so I'm still confused I guess.)

Reply:
Acquire at the memory ordering level (think atomic load operation w/ Acquire ordering).